### ISSUE TITLE
Support for dataset splits with different features 

### ISSUE THREAD

**Issue by BinyanHu** [tensorflow/datasets#977]
I am trying to implement my datasets (`DatasetsBuilders`). Different from most pre-defined datasets in tfds, these datasets have different features for each split. For example, the test set may not contain label data. The current implementation applies a uniform feature description to all splits, making it hard to fit into a dataset builder.

A reasonable solution may be modifying __init__ method of `DatasetInfo` to allow users to specify different features for each split. To be specific, it the "features" argument is a dict with `Split` as keys and `FeaturesDict` as values, it specifies different features; otherwise, if it is a single `FeaturesDict`, features for all splits are uniform (which is compatible with the current interface).

Currently, I have to write two independent builders to adapt to the difference in features.

**Comment by Conchylicultor** [comment]
Could you provide additional info for your use-case ? What would be the train/tests features that you would use ? And which dataset would this be ?
With the current implementation, the easiest way would be to create two datasets and have a higher level wrapper to switch between the two.

**Comment by BinyanHu** [comment]
I am working on some challenging datasets (PROMISE12, for example) whose labels of the test set is not open to the public. To evaluate the results on the test set, we have to submit the output predictions to the website.

In other words, the training and validation set features contain "image" and "segment", while the test set feature contains only "image".

Many datasets relating to a challenge or a competition are presented in this way, so I think it essential for tfds to support dataset splits with different features.

Thank you for your advice on creating two datasets, which, however, is not elegant for that they should be splits from the same dataset.

**Comment by Conchylicultor** [comment]
For this case, the proper way of doing this would be to have a single `FeaturesDict` for both train/test, but to have the test set return a no-op/empty value for label.

With `ClassLabel`, this is done by returning -1 for missing label.
```
yield key, {
    'image': 'path/to/image.jpg',
    'label': -1,  # Missing label
}
```

For images, currently there is no real way to have test images being optional. We could have Image returns an empty tensor for the test set if the image is missing. Something like:

```
yield key, {
    'image': 'path/to/image.jpg',
    'segment': None,  # Missing segment
}
```

**Comment by BinyanHu** [comment]
Thank you for the suggestions, but unfortunately, my datasets is a 3DImage-to-3DImage one, so to use examples with paths, a new "Feature" needs to be customized. It seems that writing two dataset builders is the best choice currently.
